modules/AlphaWalletFoundation/AlphaWalletFoundation/CoinTicker/BaseCoinTickersFetcher.swift

```diff
@@ -0,0 +1,172 @@
+//
+//  BaseCoinTickersFetcher.swift
+//  Alamofire
+//
+//  Created by Vladyslav Shepitko on 16.09.2022.
+//
+
+import Foundation
+import Combine
+import AlphaWalletCore
+
+public class BaseCoinTickersFetcher {
+    private let pricesCacheLifetime: TimeInterval = 60 * 60
+    private let dayChartHistoryCacheLifetime: TimeInterval = 60 * 60
+    private let storage: CoinTickersStorage & ChartHistoryStorage & TickerIdsStorage
+    private let networkProvider: CoinTickerNetworkProviderType
+    private let tickerIdsFetcher: TickerIdsFetcher
+    /// Cached fetch ticker prices operations
+    private var promises: AtomicDictionary<TokenMappedToTicker, AnyCancellable> = .init()
+    /// Resolving ticker ids operations
+    private var tickerResolvers: AtomicDictionary<TokenMappedToTicker, AnyCancellable> = .init()
+
+    public init(networkProvider: CoinTickerNetworkProviderType, storage: CoinTickersStorage & ChartHistoryStorage & TickerIdsStorage, tickerIdsFetcher: TickerIdsFetcher) {
+        self.networkProvider = networkProvider
+        self.tickerIdsFetcher = tickerIdsFetcher
+        self.storage = storage
+
+        //NOTE: Remove old files with tickers, ids and price histories
+        ["tickers", "tickersIds", "history"].map {
+            FileStorage().fileURL(with: $0, fileExtension: "json")
+        }.forEach { try? FileManager.default.removeItem(at: $0) }
+    }
+    public func cancel() {
+        promises.values.values.forEach { $0.cancel() }
+        promises.removeAll()
+    }
+
+    public func addOrUpdateTestsOnly(ticker: CoinTicker?, for token: TokenMappedToTicker) {
+        let tickers: [AssignedCoinTickerId: CoinTicker] = ticker.flatMap { ticker in
+            let tickerId = AssignedCoinTickerId(tickerId: "tickerId-\(token.contractAddress)-\(token.server.chainID)", token: token)
+            return [tickerId: ticker]
+        } ?? [:]
+
+        storage.addOrUpdate(tickers: tickers)
+    }
+
+    public func fetchTickers(for tokens: [TokenMappedToTicker], force: Bool = false) {
+        let targetTokensToFetchTickers = tokens.filter {
+            if promises[$0] != nil {
+                return false
+            } else {
+               return force || hasExpiredTickersLifeTimeSinceLastUpdate(for: $0)
```

Comment:
I always see the price displayed $0. Maybe the check in this line is faulty?

Reply:
oh, i see, will check it. for me it was an empty db of tickers, so might be missed this case.

Reply:
@hboon can u please share shared.realm file in dm. want to check if ticker exists

Reply:
Sent

Reply:
thanks, have u check latest pr on your side, i checked realm, it has ticker for phi native coin, so it should applear, works for me.

Reply:
No graphs appear for me.

Can you help write a Realm query that I can type in Realm Studio so I can trace which object should match? eg. `name = 'PHI'` ?

Reply:
oh you meant that there is no chart for phi token? there is no api call for retrieving prices history, so it doesn't work, ticker price only

Reply:
missed to mention this in pr description

Reply:
Oh. Ok. But my price is always $0. Maybe it’s expected when balance = 0? I forgot to check with a non-zero balance

Reply:
Ah I see what I was not finding now, we should show the current unit price for PHI (or any chain) as long as we have it, even if we don't have historical data.

From:

<img width="200" alt="Screenshot 2022-09-21 at 11 22 37 AM" src="https://user-images.githubusercontent.com/56189/191407624-8e0fbf88-3709-45d8-80c6-997ce8c9fd7f.png">

This is OK, but tap on the "Value: 3,131…" label and it should show the market price even if we don't have the historical data. Maybe just hide the appreciation/depreciation if it's not available? eg. "Market Price: 1.00 USD". Compare:

PHI's without current price|Gnosis (correct, with price)
-|-
<img width="200" alt="Screenshot 2022-09-21 at 11 22 38 AM" src="https://user-images.githubusercontent.com/56189/191407642-1b499dea-29e9-44f5-bad0-648742027482.png">|<img width="200" alt="Screenshot 2022-09-21 at 11 26 01 AM" src="https://user-images.githubusercontent.com/56189/191407647-712f33f2-7c5c-43c2-ba42-0c135672f9b0.png">

Reply:
try latest commits, its fixed there

Reply:
Ah, looks good.